src/ServiceProvider.php

```diff
@@ -104,21 +112,48 @@ protected function registerMiddleware(): void
     }
 
     /**
-     * Register data collectors and start listening for events.
+     * Register data collectors and start listening for events. Most collectors are
+     * registered by tagging the abstracts in the service container. The concreate
+     * implementations are not created during registration.
+     *
+     * An collectors which must be created prior to the boot phase should ensure
+     * they have no dependencies on other services which may not be registered yet.
+     *
+     * All tagged collectors will be gathered and given to the Agent when it is created.
      */
     protected function registerCollectors(): void
     {
-        $agent = $this->app->make(Agent::class);
-        $agent->registerCollectors();
+        if ($this->includeFrameworkEvents()) {
+            // Force the FrameworkCollector instance to be created and used. While this appears odd,
+            // the collector instance registers itself to listen for booting events, so that instance
+            // must be made available for collection later.
+            $this->app->instance(FrameworkCollector::class, $this->app->make(FrameworkCollector::class));
+
+            $this->app->tag(FrameworkCollector::class, self::COLLECTOR_TAG);
+        }
+
+        if (false !== config('elastic-apm-laravel.spans.querylog.enabled')) {
+            // DB Queries collector
+            $this->app->tag(DBQueryCollector::class, self::COLLECTOR_TAG);
+        }
+
+        // Http request collector
+        if ('cli' !== php_sapi_name()) {
```

Comment:
```suggestion
        if ($this->includeFrameworkEvents()) {
```

Reply:
Missed that. Will change.

Reply:
Addressed in new commits. I opted for a slightly different solution which I believe expresses the conditions more clearly.